### PR TITLE
external: create rados namspace for external cluster

### PR DESCRIPTION
**Pull request by parth-gr**
We need to create rados namespace if JSON data contains it Process:
1) Create the rados-namespace CR
2) Update the storage class clusterID with rados clusterID

### CODE REVIEW

File: controllers/storagecluster/storageclasses.go [hunk 1]
```diff
@@ -114,6 +114,22 @@ func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfi
 				skippedSC = append(skippedSC, sc.Name)
 				continue
 			}
+			// if rados namespace is provided, update the `sc cluster-id = rados-namespace cluster-id``
+			if strings.Contains(sc.Name, "-rados-namespace") {
+				radosNamespace := cephv1.CephBlockPoolRadosNamespace{}
+				key = types.NamespacedName{Name: radosNamspaceName, Namespace: sc.Parameters["clusterID"]}
```
Comment (Madhu-1) [inline controllers/storagecluster/storageclasses.go:120]: The namespace name is not the clusterID always isn't it?

File: controllers/storagecluster/storageclasses.go [hunk 2]
```diff
@@ -114,6 +114,22 @@ func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfi
...
+				err := r.Client.Get(context.TODO(), key, &radosNamespace)
+				if err != nil || radosNamespace.Status == nil || radosNamespace.Status.Phase != cephv1.ConditionType(util.PhaseReady) || radosNamespace.Status.Info["clusterID"] == "" {
+					r.Log.Info("Waiting for radosNamespace to be Ready. Skip reconciling StorageClass",
+						"radosNamespace", klog.KRef(key.Namespace, key.Name),
+						"it's Status %+v", radosNamespace.Status,
```
Comment (Madhu-1) [inline controllers/storagecluster/storageclasses.go:125]: i dont think we log complete Status in log

File: controllers/storagecluster/storageclasses.go [hunk 3]
```diff
@@ -114,6 +114,22 @@ func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfi
...
+						"StorageClass", klog.KRef("", sc.Name),
+					)
+					skippedSC = append(skippedSC, sc.Name)
+					continue
+				}
+				sc.Parameters["clusterID"] = radosNamespace.Status.Info["clusterID"]
```
Comment (Madhu-1) [inline controllers/storagecluster/storageclasses.go:131]: define a const for clusterID and use it

File: controllers/storagecluster/storageclasses.go [hunk 4]
```diff
@@ -88,7 +88,7 @@ func (obj *ocsStorageClass) ensureDeleted(r *StorageClusterReconciler, instance
 	return reconcile.Result{}, nil
 }
 
-func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfiguration) error {
+func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfiguration, radosNamspaceName string) error {
```
Comment (Madhu-1) [inline controllers/storagecluster/storageclasses.go:91]: i think we dont need to pass the radosNamepsace name as a function parameter, just use the const name as it used for other resouces

File: controllers/storagecluster/storageclasses.go [hunk 5]
```diff
@@ -114,6 +114,22 @@ func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfi
 				skippedSC = append(skippedSC, sc.Name)
 				continue
 			}
+			// if rados namespace is provided, update the `sc cluster-id = rados-namespace cluster-id``
+			if strings.Contains(sc.Name, "-rados-namespace") {
```
Comment (Madhu-1) [inline controllers/storagecluster/storageclasses.go:118]: also check if ceph cluster is external as well

File: controllers/storagecluster/external_resources.go [hunk 6]
```diff
@@ -344,12 +348,28 @@ func (r *StorageClusterReconciler) createExternalStorageClusterResources(instanc
...
+			if d.Name == radosNamespace {
+				radosNamespaceName = d.Data["radosNamespace"]
+				radosNamespace := &cephv1.CephBlockPoolRadosNamespace{ObjectMeta: objectMeta}
+				mutateFn := func() error {
+					radosNamespace.Spec = cephv1.CephBlockPoolRadosNamespaceSpec{
+						BlockPoolName: d.Data["radosNamespace"],
+					}
+					return nil
+				}
+				_, err := ctrl.CreateOrUpdate(context.TODO(), r.Client, radosNamespace, mutateFn)
```
Comment (Madhu-1) [inline controllers/storagecluster/external_resources.go:361]: where are we taking care of deletion of rados namespace?

File: controllers/storagecluster/external_resources.go [hunk 7]
```diff
@@ -344,12 +348,28 @@ func (r *StorageClusterReconciler) createExternalStorageClusterResources(instanc
 				r.Log.Error(err, "Could not create CephFilesystemSubVolumeGroup.", "CephFilesystemSubVolumeGroup", klog.KRef(found.Namespace, found.Name))
 				return err
 			}
+		case "Namespace":
```
Comment (Madhu-1) [inline controllers/storagecluster/external_resources.go:351]: This should be `CephBlockPoolRadosNamespace` ? and i don't see this spec passed in the Rook PR.

File: controllers/storagecluster/external_resources.go [hunk 8]
```diff
@@ -344,12 +348,28 @@ func (r *StorageClusterReconciler) createExternalStorageClusterResources(instanc
...
+			if d.Name == radosNamespace {
+				radosNamespaceName = d.Data["radosNamespace"]
+				radosNamespace := &cephv1.CephBlockPoolRadosNamespace{ObjectMeta: objectMeta}
+				mutateFn := func() error {
+					radosNamespace.Spec = cephv1.CephBlockPoolRadosNamespaceSpec{
+						BlockPoolName: d.Data["radosNamespace"],
```
Comment (Madhu-1) [inline controllers/storagecluster/external_resources.go:357]: This should be the blockpool name not the rados namespace name

File: controllers/storagecluster/external_resources.go [hunk 9]
```diff
@@ -344,12 +348,28 @@ func (r *StorageClusterReconciler) createExternalStorageClusterResources(instanc
 				r.Log.Error(err, "Could not create CephFilesystemSubVolumeGroup.", "CephFilesystemSubVolumeGroup", klog.KRef(found.Namespace, found.Name))
 				return err
 			}
+		case "Namespace":
+			if d.Name == radosNamespace {
+				radosNamespaceName = d.Data["radosNamespace"]
```
Comment (Madhu-1) [inline controllers/storagecluster/external_resources.go:353]: why do we need this one?

**Review by Madhu-1 — COMMENTED** [review]
Add Unit testing for the new SC?

File: controllers/storagecluster/storageclasses.go (hunk 2 above)
Comment (parth-gr) [reply controllers/storagecluster/storageclasses.go:125]: It's need to know why it fails

File: controllers/storagecluster/storageclasses.go (hunk 3 above)
Comment (parth-gr) [reply controllers/storagecluster/storageclasses.go:131]: updated a new namespace parmater now now needed

File: controllers/storagecluster/external_resources.go (hunk 6 above)
Comment (parth-gr) [reply controllers/storagecluster/external_resources.go:361]: where it should be present, with blockpool deletion?

Comment (parth-gr) [reply controllers/storagecluster/external_resources.go:361]: @iamniting can you help with ths telling the right pace for the external cluster reconcile

File: controllers/storagecluster/storageclasses.go [hunk 10]
```diff
@@ -43,7 +43,7 @@ func (obj *ocsStorageClass) ensureCreated(r *StorageClusterReconciler, instance
 		return reconcile.Result{}, err
 	}
 
-	err = r.createStorageClasses(scs)
+	err = r.createStorageClasses(scs, "", instance.Namespace)
```
Comment (umangachapagain) [inline controllers/storagecluster/storageclasses.go:46]: Please add a comment about why we are passing an empty string. Makes it easier to understand.

File: controllers/storagecluster/storageclasses.go [hunk 11]
```diff
@@ -104,7 +104,7 @@ func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfi
 		case (strings.Contains(sc.Name, "-ceph-rbd") || strings.Contains(sc.Provisioner, fmt.Sprintf("%s.rbd.csi.ceph.com", operatorNamespace))) && !scc.isClusterExternal:
 			// wait for CephBlockPool to be ready
 			cephBlockPool := cephv1.CephBlockPool{}
-			key := types.NamespacedName{Name: sc.Parameters["pool"], Namespace: sc.Parameters["clusterID"]}
+			key := types.NamespacedName{Name: sc.Parameters["pool"], Namespace: namespace}
```
Comment (umangachapagain) [inline controllers/storagecluster/storageclasses.go:107]: Is this change related to creating rados namespace?

File: controllers/storagecluster/storageclasses.go [hunk 12]
```diff
@@ -249,14 +267,17 @@ func newCephFilesystemStorageClassConfiguration(initData *ocsv1.StorageCluster)
 }
 
 // newCephBlockPoolStorageClassConfiguration generates configuration options for a Ceph Block Pool StorageClass.
-func newCephBlockPoolStorageClassConfiguration(initData *ocsv1.StorageCluster) StorageClassConfiguration {
+func newCephBlockPoolStorageClassConfiguration(initData *ocsv1.StorageCluster, storageClassName string) StorageClassConfiguration {
```
Comment (umangachapagain) [inline controllers/storagecluster/storageclasses.go:270]: Create another function for radosStorageClassConfiguration() that takes StorageClass name as an arg and calls this function for better readability. Do not change existing functions to do multiple things.

File: controllers/storagecluster/storageclasses.go (hunk 2 above)
Comment (umangachapagain) [reply controllers/storagecluster/storageclasses.go]: It's always available in the CR and we collect it separately. Let's not log this.

File: controllers/storagecluster/storageclasses.go (hunk 4 above)
Comment (umangachapagain) [reply controllers/storagecluster/storageclasses.go]: If it's not needed now, I am in favor of using constant rather than changing function signature. We can always change it later.

File: controllers/storagecluster/storageclasses.go (hunk 2 above)
Comment (parth-gr) [reply controllers/storagecluster/storageclasses.go]: Updated

File: controllers/storagecluster/storageclasses.go (hunk 4 above)
Comment (parth-gr) [reply controllers/storagecluster/storageclasses.go]: updated

File: controllers/storagecluster/external_resources.go [hunk 13]
```diff
@@ -344,13 +347,31 @@ func (r *StorageClusterReconciler) createExternalStorageClusterResources(instanc
 				r.Log.Error(err, "Could not create CephFilesystemSubVolumeGroup.", "CephFilesystemSubVolumeGroup", klog.KRef(found.Namespace, found.Name))
 				return err
 			}
+		case "CephBlockPoolRadosNamespace":
+			radosNamespace = d.Data["radosNamespace"]
+			rbdPool := d.Data["pool"]
+			radosNamespace := &cephv1.CephBlockPoolRadosNamespace{ObjectMeta: objectMeta}
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply controllers/storagecluster/external_resources.go:353]: Corrected, actually by-mistake I renamed the variable as per the suggestion

File: controllers/storagecluster/storageclasses.go (hunk 10 above)
Comment (parth-gr) [reply controllers/storagecluster/storageclasses.go:46]: removed

File: controllers/storagecluster/storageclasses.go (hunk 11 above)
Comment (parth-gr) [reply controllers/storagecluster/storageclasses.go:107]: Yes, the `sc.Parameters["clusterID"]` will be updated with new value if rados namespace is passed.
So the actual namespace would be different

File: controllers/storagecluster/external_resources.go (hunk 6 above)
Comment (parth-gr) [reply controllers/storagecluster/external_resources.go]: I see we do expect that services created by external cluster should be deleted by its own reconcile https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/storagecluster/external_resources.go#L274

So I think the better place would be with cephblockpool reconcile

Comment (iamniting) [reply controllers/storagecluster/external_resources.go]: Lets add a owner reference that would be enough.

Comment (parth-gr) [reply controllers/storagecluster/external_resources.go]: We are good with it as we have already mentioned the objectmeta

**Comment by parth-gr** [comment]
@umangachapagain any other thoughts?

**Comment by parth-gr** [comment]
Getting 
```
W1127 14:04:23.595231       1 reflector.go:535] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:208: failed to list *v1.CephBlockPoolRadosNamespace: cephblockpoolradosnamespaces.ceph.rook.io is forbidden: User "system:serviceaccount:openshift-storage:ocs-operator" cannot list resource "cephblockpoolradosnamespaces" in API group "ceph.rook.io" in the namespace "openshift-storage"
E1127 14:04:23.595413       1 reflector.go:147] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:208: Failed to watch *v1.CephBlockPoolRadosNamespace: failed to list *v1.CephBlockPoolRadosNamespace: cephblockpoolradosnamespaces.ceph.rook.io is forbidden: User "system:serviceaccount:openshift-storage:ocs-operator" cannot list resource "cephblockpoolradosnamespaces" in API group "ceph.rook.io" in the namespace "openshift-storage"

```

@iamniting is look like rbac issue?

**Comment by iamniting** [comment]
> Getting
> 
> ```
> W1127 14:04:23.595231       1 reflector.go:535] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:208: failed to list *v1.CephBlockPoolRadosNamespace: cephblockpoolradosnamespaces.ceph.rook.io is forbidden: User "system:serviceaccount:openshift-storage:ocs-operator" cannot list resource "cephblockpoolradosnamespaces" in API group "ceph.rook.io" in the namespace "openshift-storage"
> E1127 14:04:23.595413       1 reflector.go:147] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:208: Failed to watch *v1.CephBlockPoolRadosNamespace: failed to list *v1.CephBlockPoolRadosNamespace: cephblockpoolradosnamespaces.ceph.rook.io is forbidden: User "system:serviceaccount:openshift-storage:ocs-operator" cannot list resource "cephblockpoolradosnamespaces" in API group "ceph.rook.io" in the namespace "openshift-storage"
> ```
> 
> @iamniting is look like rbac issue?

You need to add permissions here https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/storagecluster/reconcile.go#L130

File: config/rbac/role.yaml [hunk 14]
```diff
@@ -39,17 +39,7 @@ rules:
...
-  - create
-  - delete
-  - get
-  - list
-  - update
-  - watch
-- apiGroups:
-  - ceph.rook.io
-  resources:
+  - CephBlockPoolRadosNamespace
```
Comment (BlaineEXE) [inline config/rbac/role.yaml:42]: Nit: this could be `cephblockpoolradosnamespaces` to keep the capitalization and pluralization consistent with the rest of the resources.

File: controllers/storagecluster/reconcile.go [hunk 15]
```diff
@@ -107,7 +107,7 @@ var validTopologyLabelKeys = []string{
 }
 
 // +kubebuilder:rbac:groups=ocs.openshift.io,resources=*,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=ceph.rook.io,resources=cephclusters;cephblockpools;cephfilesystems;cephnfses;cephobjectstores;cephobjectstoreusers;cephrbdmirrors,verbs=*
+// +kubebuilder:rbac:groups=ceph.rook.io,resources=cephclusters;cephblockpools;cephfilesystems;cephnfses;cephobjectstores;cephobjectstoreusers;cephrbdmirrors;CephBlockPoolRadosNamespace,verbs=*
```
Comment (BlaineEXE) [inline controllers/storagecluster/reconcile.go:110]: Extra empty line necessary?

File: deploy/csv-templates/ocs-operator.csv.yaml.in [hunk 16]
```diff
@@ -243,17 +243,7 @@ spec:
...
-          - create
-          - delete
-          - get
-          - list
-          - update
-          - watch
-        - apiGroups:
-          - ceph.rook.io
-          resources:
+          - CephBlockPoolRadosNamespace
```
Comment (BlaineEXE) [inline deploy/csv-templates/ocs-operator.csv.yaml.in:246]:
```suggestion
          - cephblockpoolradosnamespaces
```

**Review by BlaineEXE — COMMENTED** [review]
Just a few very minor thoughts.

File: config/rbac/role.yaml (hunk 14 above)
Comment (parth-gr) [reply config/rbac/role.yaml]: Updated

File: controllers/storagecluster/storageclasses.go (hunk 11 above)
Comment (umangachapagain) [reply controllers/storagecluster/storageclasses.go:109]: What happens to existing StorageClasses? Will it affect upgrades?

Comment (parth-gr) [reply controllers/storagecluster/storageclasses.go:109]: No it wont effect with upgrades, and with new clusters either one of it will get creating according to the paramters

File: controllers/storagecluster/reconcile.go [hunk 17]
```diff
@@ -107,7 +107,7 @@ var validTopologyLabelKeys = []string{
 }
 
 // +kubebuilder:rbac:groups=ocs.openshift.io,resources=*,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=ceph.rook.io,resources=cephclusters;cephblockpools;cephfilesystems;cephnfses;cephobjectstores;cephobjectstoreusers;cephrbdmirrors,verbs=*
+// +kubebuilder:rbac:groups=ceph.rook.io,resources=cephclusters;cephblockpools;cephfilesystems;cephnfses;cephobjectstores;cephobjectstoreusers;cephrbdmirrors;cephBlockPoolRadosNamespace,verbs=*
```
Comment (BlaineEXE) [inline controllers/storagecluster/reconcile.go:110]: Code changes should try to keep new items formatted in the same way as earlier items. This is stil in cAmElCaP, but everything else is allinlowercase. Also, this is not plural, where all other items are. Please fix both, like so.

```suggestion
// +kubebuilder:rbac:groups=ceph.rook.io,resources=cephclusters;cephblockpools;cephfilesystems;cephnfses;cephobjectstores;cephobjectstoreusers;cephrbdmirrors;cephblockpoolradosnamespaces,verbs=*
```

File: deploy/csv-templates/ocs-operator.csv.yaml.in [hunk 18]
```diff
@@ -243,17 +243,7 @@ spec:
...
-          - create
-          - delete
-          - get
-          - list
-          - update
-          - watch
-        - apiGroups:
-          - ceph.rook.io
-          resources:
+          - cephBlockPoolRadosNamespace
```
Comment (BlaineEXE) [inline deploy/csv-templates/ocs-operator.csv.yaml.in:246]:
```suggestion
          - cephblockpoolradosnamespaces
```

File: deploy/ocs-operator/manifests/ocs-operator.clusterserviceversion.yaml [hunk 19]
```diff
@@ -1973,17 +1973,7 @@ spec:
...
-          - create
-          - delete
-          - get
-          - list
-          - update
-          - watch
-        - apiGroups:
-          - ceph.rook.io
-          resources:
+          - cephBlockPoolRadosNamespace
```
Comment (BlaineEXE) [inline deploy/ocs-operator/manifests/ocs-operator.clusterserviceversion.yaml:1976]:
```suggestion
          - cephblockpoolradosnamespaces
```

**Review by BlaineEXE — COMMENTED** [review]
I think one of my comments was unclear. The new resource being added is still not formatted to match the existing code. People tend to have better list comprehension when things are formatted similarly.

File: config/rbac/role.yaml [hunk 20]
```diff
@@ -39,17 +39,7 @@ rules:
...
-  - create
-  - delete
-  - get
-  - list
-  - update
-  - watch
-- apiGroups:
-  - ceph.rook.io
-  resources:
+  - cephBlockPoolRadosNamespace
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply config/rbac/role.yaml]: Updated
